cpp/include/cudf/scalar/scalar.hpp

```diff
@@ -0,0 +1,182 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/types.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_scalar.hpp>
+#include <rmm/thrust_rmm_allocator.h>
+#include <rmm/mr/device_memory_resource.hpp>
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+// Forward declarations
+template <typename T>
+class numeric_scalar_device_view;
+
+class string_scalar_device_view;
+
+template <typename T>
+class timestamp_scalar_device_view;
+
+namespace cudf {
+
+class scalar {
+ public:
+  scalar() = default;
+  ~scalar() = default;
+  scalar& operator=(scalar const& other) = delete;
+  scalar& operator=(scalar&& other) = delete;
+  
+  scalar(data_type type, bool is_valid = false, cudaStream_t stream_ = 0,
+      rmm::mr::device_memory_resource *mr_ = rmm::mr::get_default_resource())
+   : _type(type), _is_valid(is_valid, stream_, mr_)
+  {}
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new scalar by deep copying the contents of `other`.
+   *
+   * All device memory allocation and copying is done using the
+   * `device_memory_resource` and `stream` from `other`.
+   *
+   * @param other The scalar to copy
+   *---------------------------------------------------------------------------**/
+  scalar(scalar const& other);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Move the contents from `other` to create a new scalar.
+   *
+   * After the move, `other.type() = {EMPTY}`
+   *
+   * @param other The scalar whose contents will be moved into the new scalar
+   *---------------------------------------------------------------------------**/
+  scalar(scalar&& other);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns the scalar's logical element type
+   *---------------------------------------------------------------------------**/
+  data_type type() const noexcept { return _type; }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Sets this scalar to null
+   *---------------------------------------------------------------------------**/
+  void set_null() { _is_valid.set_value(false); }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Sets this scalar to valid
+   *---------------------------------------------------------------------------**/
+  void set_valid() { _is_valid.set_value(true); }
```

Comment:
Why not remove `set_null()` and have one function `void set_valid(bool is_valid) { _is_valid.set_value(is_valid); }`?  This would be more consistent with the single `is_valid()` function.